CLDSRV-488: fix website 404 with bucket policy

BourgoisMickael opened the pull request:
If object doesn’t exist but bucket policy allows access to the object, the response should be a 404 and not a 403.

BourgoisMickael committed:
CLDSRV-488: fix website 404 with bucket policy

If bucket policy authorize access to a
non existant object, there should be a 404
and not a 403

BourgoisMickael committed:
CLDSRV-488: test website 404 with bucket policy

bert-e commented:
# Hello bourgoismickael,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).





*Status report is not available.*

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

BourgoisMickael commented:
/create_integration_branches

bert-e commented:
# Unknown command



I didn't understand this comment by @bourgoismickael:

> /create_integration_branch

I don't know what `create_integration_branch` means.

Please **edit** or **delete** the corresponding comment so I can move on.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/CLDSRV-488-error-type-bp` into
`development/7.10`
* [w/7.70/bugfix/CLDSRV-488-error-type-bp](https://github.com/scality/cloudserver/compare/development/7.70...w/7.70/bugfix/CLDSRV-488-error-type-bp) will be merged into `development/7.70` 
* [w/8.6/bugfix/CLDSRV-488-error-type-bp](https://github.com/scality/cloudserver/compare/development/8.6...w/8.6/bugfix/CLDSRV-488-error-type-bp) will be merged into `development/8.6` 
* [w/8.7/bugfix/CLDSRV-488-error-type-bp](https://github.com/scality/cloudserver/compare/development/8.7...w/8.7/bugfix/CLDSRV-488-error-type-bp) will be merged into `development/8.7` 
* [w/8.8/bugfix/CLDSRV-488-error-type-bp](https://github.com/scality/cloudserver/compare/development/8.8...w/8.8/bugfix/CLDSRV-488-error-type-bp) will be merged into `development/8.8` 



The following branches will **NOT** be impacted:

* `development/7.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

jonathan-gramain reviewed (commented):
Maybe it technically works but I'm not fond of this approach, I see potentially two issues:
- changing `requestType` may cause unwanted side-effects down the line, during or after `checkBucketPolicy` is done. At least if it is a workaround, we should revert it to its original value when we don't need the change anymore, and add a clear comment why we change it in the first place
- I think it deserves additional comments explaining more in depth the reasoning, it's not clear to me and I don't know where to look to find the explanation of why the `objectGet` request type must be set for bucket website policy checks.

Also if there's a cleaner way to deal with this issue that doesn't involve a special flag in the request object nor to change the request type, I'd prefer that.

williamlardier commented on lib/api/websiteGet.js:
```diff
@@ -138,6 +138,10 @@ function websiteGet(request, log, callback) {
                     log.trace('error retrieving object metadata',
                     { error: err });
                     let returnErr = err;
+                    // requestType must be bucketGet to check bucket acl
+                    // but objectGet to check bucket policy
+                    // eslint-disable-next-line no-param-reassign
+                    request.websiteOverrideRequestTypeForBucketPolicy = true;
```
As this is API-specific logic, would it work better if we handle the return code directly in the API, in case of an access denied?
Based on the authz results from IAM,overriding the request type may cause corner cases where some actions are allowed or not by the ACLs and bucket policy. I think, if we need to check both bucket ACLs and ObjectGet permission, that we should call this function twice and behave like AWS.

In websiteGet, we already check the bucket ACLs and policy with the `bucketGet` request type in case we have an error getting the object MD:
```js
// if index object does not exist and bucket is private AWS
                    // returns 403 - AccessDenied error.
                    if (err.is.NoSuchKey && !bucketAuthorized) {
                        returnErr = errors.AccessDenied;
                    }
                    return _errorActions(returnErr,
                      websiteConfig.getErrorDocument(), routingRules,
                      bucket, reqObjectKey, corsHeaders, request, log,
                      callback);
```
Here, we should be able to either evaluate the `objectGet` permission to return 404. We have the standardMetadataValidateBucketAndObj function that can be used, and from which you can override the request type, without altering the request object, eg:

```js
    const mdValParams = {
        authInfo,
        bucketName,
        objectKey,
        versionId,
        getDeleteMarker: true,
        requestType: 'objectHead',
        request,
    };

    return standardMetadataValidateBucketAndObj(mdValParams, request.actionImplicitDenies, log, [...]
```

BourgoisMickael replied:
I can't run it twice, all these functions seems to encapsulate both `checkBucketAcls` and `processBucketPolicy` but I need to run them separately for `bucketGet` and `objectGet`.

`processBucketPolicy` with `bucketGet` is invalid because website only accepts `s3:GetObject` so if there is `s3:ListBucket` in a policy this will result in an invalid access.

`checkBucketAcls` returns always `true` for `objectGet` (to delegate to object acl, but I'm in the case where there is no object to check later so I should not do this check)

williamlardier replied:
The idea is to have an additional permission check, only if the object is not found but we are supposed to have access to it (given that it exists), in this case, we should check the policies against `s3:ObjectGet` instead of `s3:bucketGet`.
If the `isBucketAuthorized` or `isObjAuthorized` are checking the ACLs but you want to skip them, shouldn't we just update these functions with a new optional parameter so that we skip the ACL checks, to avoid changing the generic function?

BourgoisMickael replied:
Yes but I didn't wanted to add another parameter as theses functions already have a lot of parameters.
But I can do that so I don't put a flag on the request and I can override the request type earlier

BourgoisMickael commented:
> Maybe it technically works but I'm not fond of this approach, I see potentially two issues:
> 
> * changing `requestType` may cause unwanted side-effects down the line, during or after `checkBucketPolicy` is done. At least if it is a workaround, we should revert it to its original value when we don't need the change anymore, and add a clear comment why we change it in the first place
> * I think it deserves additional comments explaining more in depth the reasoning, it's not clear to me and I don't know where to look to find the explanation of why the `objectGet` request type must be set for bucket website policy checks.
> 
> Also if there's a cleaner way to deal with this issue that doesn't involve a special flag in the request object nor to change the request type, I'd prefer that.

@jonathan-gramain 

The website doesn't work exactly like a `GetObject`.
`GetObject` needs `s3:ListBucket` action on bucket to determine if an error is 404 or 403.

But website only uses `s3:GetObject` in bucket policy: https://docs.aws.amazon.com/AmazonS3/latest/userguide/WebsiteAccessPermissionsReqd.html#bucket-policy-static-site

They don't mention `s3:ListBucket` and after doing some tests, it doesn't work at all.

So an `s3:ListBucket` in bucket policy is ignored by S3 and only `s3:GetObject` can be used to determine if an error is 403 or 404.

Now if I use this policy:

```json
{
   "Version": "2012-10-17",
   "Statement": [
      {
         "Sid": "PublicListObject",
         "Effect": "Allow",
         "Principal": "*",
         "Action": [
            "s3:ListBucket",
            "s3:PutObjectTagging"
         ],
         "Resource": [
            "arn:aws:s3:::mickael-bourgois-err",
            "arn:aws:s3:::mickael-bourgois-err/*"
         ]
      }
   ]
}
```
Cloudserver in case of error, with the `bucketGet` will apply the `s3:ListBucket` (that is a bucket action and not an object action) to any object because of the resource with `/*`. (I think it should not do that btw) and return 404s instead of 403.

The expected behavior on S3 would be:
```
curl --head http://mickael-bourgois-err.s3-website.eu-north-1.amazonaws.com/invalid
HTTP/1.1 403 Forbidden
x-amz-error-code: AccessDenied
x-amz-error-message: Access Denied
x-amz-request-id: 90MB6HKNG7VAKEWK
x-amz-id-2: 7FYMC9Iv8Idc0RWPEWtMXFpVV2aDETNzTzRzXNdSy2L/NmayXsBvQ+d0DNFfH4cZ1HGeElYekBI=
Date: Tue, 09 Jan 2024 13:41:24 GMT
Server: AmazonS3
```
___

We can't replace `bucketGet` with `objectGet` because `checkBucketAcls` will return `true`.

That's why we need
- `checkBucketAcls` to use `bucketGet`
- `checkBucketPolicy` to use `objectGet`
___

I moved the override to `isBucketAuthorized` and replaced the flag with an new optional argument

BourgoisMickael committed:
CLDSRV-488: move website condition, replace flag

bert-e commented:
# Conflict




There is a conflict between your branch `bugfix/CLDSRV-488-error-type-bp` and the
destination branch `development/7.10`.

Please resolve the conflict on **the feature branch** (`bugfix/CLDSRV-488-error-type-bp`).

```sh
git fetch && \
git checkout origin/bugfix/CLDSRV-488-error-type-bp && \
git merge origin/development/7.10
```

Resolve merge conflicts and commit

```sh
git push origin HEAD:bugfix/CLDSRV-488-error-type-bp
```







*The following options are set:* **create_integration_branches**

BourgoisMickael committed:
Merge branch 'development/7.10' into bugfix/CLDSRV-488-error-type-bp

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

jonathan-gramain commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -346,6 +346,15 @@ function isBucketAuthorized(bucket, requestTypesInput, canonicalID, authInfo, lo
+        if (isWebsite && _requestType === 'bucketGet') {
+            // eslint-disable-next-line no-param-reassign
+            _requestType = 'objectGet';
+            actionImplicitDenies[_requestType] = actionImplicitDenies[_requestType] || false;
```
minor clarity suggestion

```suggestion
            actionImplicitDenies['objectGet'] = actionImplicitDenies['objectGet'] || false;
```

jonathan-gramain reviewed (approved):
I think it is better this way, and clearer with the comment, thanks

BourgoisMickael committed:
CLDSRV-488: apply review suggestion

Co-authored-by: Jonathan Gramain <jonathan.gramain@scality.com>

BourgoisMickael commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_integration_branches**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/CLDSRV-488-error-type-bp` into
`development/7.10`
* [w/7.70/bugfix/CLDSRV-488-error-type-bp](https://github.com/scality/cloudserver/compare/development/7.70...w/7.70/bugfix/CLDSRV-488-error-type-bp) will be merged into `development/7.70` 
* [w/8.6/bugfix/CLDSRV-488-error-type-bp](https://github.com/scality/cloudserver/compare/development/8.6...w/8.6/bugfix/CLDSRV-488-error-type-bp) will be merged into `development/8.6` 
* [w/8.7/bugfix/CLDSRV-488-error-type-bp](https://github.com/scality/cloudserver/compare/development/8.7...w/8.7/bugfix/CLDSRV-488-error-type-bp) will be merged into `development/8.7` 
* [w/8.8/bugfix/CLDSRV-488-error-type-bp](https://github.com/scality/cloudserver/compare/development/8.8...w/8.8/bugfix/CLDSRV-488-error-type-bp) will be merged into `development/8.8` 



The following branches will **NOT** be impacted:

* `development/7.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

BourgoisMickael committed:
CLDSRV-488: fix lint

BourgoisMickael commented:
/approve

williamlardier commented on lib/api/apiUtils/authorization/permissionChecks.js:
```diff
@@ -322,7 +322,7 @@ function processBucketPolicy(requestType, bucket, canonicalID, arn, bucketOwner,
 }
 
 function isBucketAuthorized(bucket, requestTypesInput, canonicalID, authInfo, log, request,
-    actionImplicitDeniesInput = {}) {
+    actionImplicitDeniesInput = {}, isWebsite = false) {
```
Echoing your comment around the number of args here, another way could be to have a last parameter as an object, where we can set various properties depending on the needs, it would allow us to reuse if we need more specific logic, but not a blocker for this PR

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/cloudserver/commit/621cb3368018ce71b3a99ffdc2f6289a8ab037e3) did not succeed in branch w/8.6/bugfix/CLDSRV-488-error-type-bp.




*The following options are set:* **approve, create_integration_branches**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/cloudserver/commit/a4cc5e45f386a85dd7f45b82ba8be9c0f300bb97) did not succeed in branch w/8.7/bugfix/CLDSRV-488-error-type-bp.




*The following options are set:* **approve, create_integration_branches**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/7.10`

* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`

* :heavy_check_mark: `development/8.8`



The following branches will **NOT** be impacted:

* `development/7.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve, create_integration_branches**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/7.10`

* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`

* :heavy_check_mark: `development/8.8`



The following branches have **NOT** changed:

* `development/7.4`



Please check the status of the associated issue CLDSRV-488.

Goodbye bourgoismickael.
